vpn/src/main/java/com/cliqz/browser/vpn/core/ExtAuthHelper.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2012-2018 Arne Schwabe
+ * Distributed under the GNU GPL v2 with additional terms. For full terms see the file doc/LICENSE.txt
+ */
+
+package com.cliqz.browser.vpn.core;
+
+import android.annotation.TargetApi;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.ServiceConnection;
+import android.content.pm.PackageManager;
+import android.content.pm.ResolveInfo;
+import android.os.*;
+import android.security.KeyChainException;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.WorkerThread;
+import android.widget.ArrayAdapter;
+import android.widget.Spinner;
+import android.widget.SpinnerAdapter;
+import com.cliqz.browser.vpn.api.ExternalCertificateProvider;
+
+import java.io.ByteArrayInputStream;
+import java.io.Closeable;
+import java.io.UnsupportedEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.*;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class ExtAuthHelper {
+
+    public static final String ACTION_CERT_CONFIGURATION = "com.cliqz.browser.vpn.api.ExternalCertificateConfiguration";
+    public static final String ACTION_CERT_PROVIDER = "com.cliqz.browser.vpn.api.ExternalCertificateProvider";
+
+    public static final String EXTRA_ALIAS = "com.cliqz.browser.vpn.api.KEY_ALIAS";
+    public static final String EXTRA_DESCRIPTION = "com.cliqz.browser.vpn.api.KEY_DESCRIPTION";
+
+
+    public static void setExternalAuthProviderSpinnerList(Spinner spinner, String selectedApp) {
+        Context c = spinner.getContext();
+        final PackageManager pm = c.getPackageManager();
+        ArrayList<ExternalAuthProvider> extProviders = getExternalAuthProviderList(c);
+
+        int selectedPos = -1;
+
+        if (extProviders.size() ==0)
+        {
+            selectedApp = "";
+            ExternalAuthProvider noauthprovider = new ExternalAuthProvider();
+            noauthprovider.label = "No external auth provider found";
+            noauthprovider.packageName = selectedApp;
+            noauthprovider.configurable = false;
+            extProviders.add(noauthprovider);
+        }
+
+
+        for (int i = 0; i < extProviders.size(); i++) {
+            if (extProviders.get(i).packageName.equals(selectedApp))
+                selectedPos = i;
+        }
+        SpinnerAdapter extAppAdapter = new ArrayAdapter<ExternalAuthProvider>(c, android.R.layout.simple_spinner_item, android.R.id.text1, extProviders);
+        spinner.setAdapter(extAppAdapter);
+        if (selectedPos != -1)
+            spinner.setSelection(selectedPos);
+    }
+
+    static ArrayList<ExternalAuthProvider> getExternalAuthProviderList(Context c) {
```

Comment:
Can return `List<ExternalAuthProvider>`. Programming to interface not implementation.

http://thefinestartist.com/effective-java/52

Reply:
All the code inside vpn module is not mine.